src/runtime/server/lib/oauth/bluesky.ts

```diff
@@ -0,0 +1,201 @@
+import type { H3Event } from 'h3'
+import { createError, eventHandler, getQuery, sendRedirect } from 'h3'
+import type { Storage, StorageValue } from 'unstorage'
+import { NodeOAuthClient, OAuthCallbackError, OAuthResolverError, OAuthResponseError } from '@atproto/oauth-client-node'
+import type {
+  NodeSavedSession,
+  NodeSavedSessionStore,
+  NodeSavedState,
+  NodeSavedStateStore,
+} from '@atproto/oauth-client-node'
+import { defu } from 'defu'
+import { Agent } from '@atproto/api'
+import type { AppBskyActorGetProfile } from '@atproto/api'
+import { getOAuthRedirectURL, handleMissingConfiguration } from '../utils'
+import type { OAuthConfig } from '#auth-utils'
+import { useRuntimeConfig, useStorage } from '#imports'
+
+export interface OAuthBlueskyConfig {
+  /**
+   * The URL on which your app will be deployed. This will be used to enable Bluesky to validate the client.
+   * This is only required for production environments.
+   * @default process.env.NUXT_OAUTH_BLUESKY_PUBLIC_URL
+   */
+  publicUrl?: string
+  /**
+   * Redirect URL to to allow overriding for situations like prod failing to determine public hostname
+   * @default process.env.NUXT_OAUTH_BLUESKY_REDIRECT_URL
+   */
+  redirectURL?: string
+  /**
+   * Bluesky OAuth Scope. The `atproto` scope is required and will be added if not present.
+   * @default ['atproto']
+   * @see https://atproto.com/specs/oauth#authorization-scopes
+   * @example ['atproto', 'transition:generic']
+   */
+  scope?: string[]
+  /**
+   * Human-readable name of the client.
+   */
+  clientName?: string
+  /**
+   *  This is a homepage URL for the client. If provided, the client_uri must have the same hostname as client_id.
+   */
+  clientUri?: string
+}
+
+type BlueSkyUser = AppBskyActorGetProfile.Response['data'] | null
+type BlueSkyTokens = NodeSavedSession['tokenSet']
+
+export function defineOAuthBlueskyEventHandler({ config, onSuccess, onError }: OAuthConfig<OAuthBlueskyConfig, BlueSkyUser, BlueSkyTokens>) {
+  return eventHandler(async (event: H3Event) => {
+    config = defu(config, useRuntimeConfig(event).oauth?.bluesky) as OAuthBlueskyConfig
+
+    config.scope ||= []
+    if (!config.scope.includes('atproto')) {
+      // atproto is a required scope
+      config.scope.push('atproto')
+    }
+
+    const scope = config.scope.join(' ')
```

Comment:
```suggestion
    const scope = [...new Set(config.scope)].scope.join(' ')
```

Reply:
I don't think this is the way, as of now it's currently causing a memory leak as every request will add the configured scope to the list of scopes. I should patch the merge strategy of `defu` instead